File: Packages/StrandAnalytics/Sources/StrandAnalytics/DailyEffortScorer.swift
import Foundation
import WhoopProtocol

/// Resolves one daily Effort value from cardiovascular load and low-intensity movement.
///
/// Edwards TRIMP intentionally contributes nothing below 50% heart-rate reserve. That is useful for
/// workout load, but it makes an otherwise active walking day read as zero. This kernel adds a
/// conservative movement floor without adding it to cardiovascular load:
///
/// - measured/calibrated steps are converted to walking minutes at 100 steps/minute;
/// - each walking minute receives one quarter of Edwards zone 1's weight;
/// - the resulting low-intensity TRIMP uses the same logarithmic 0...100 map as cardiovascular Effort;
/// - the final score is `max(cardio, movement)`, so workout steps are not counted twice.
///
/// When a step counter is unavailable, dense-enough wrist gravity can provide a conservative fallback.
/// Missing movement leaves the cardiovascular score byte-identical. This is an approximate activity-load
/// model, not a reproduction of any proprietary wearable score.
public enum DailyEffortScorer {
    /// Common cadence reference for purposeful walking.
    public static let walkingCadenceStepsPerMinute: Double = 100
    /// Sub-zone movement gets one quarter of the lowest Edwards zone weight.
    public static let lightMovementWeight: Double = 0.25
    /// Sustained walking is approximately 0.2...0.4 g on the existing NOOP wrist-motion feature.
    public static let walkingMotionThresholdG: Double = 0.15
    /// Do not infer activity across a telemetry hole longer than two minutes.
    public static let maximumMotionEvidenceGapSeconds: Double = 120
    /// One motion observation may represent at most one minute of activity.
    public static let maximumCreditedMotionSeconds: Double = 60

    /// Resolve daily Effort. A real movement score may stand alone when HR coverage is insufficient.
    public static func score(
        cardioEffort: Double?,
        steps: Int?,
        gravity: [GravitySample] = []
    ) -> Double? {
        guard let movement = movementEffort(steps: steps, gravity: gravity) else {
            return cardioEffort
        }
        guard let cardioEffort else { return movement }
        return max(cardioEffort, movement)
    }

    /// Prefer calibrated steps; fall back to measured wrist movement only when steps are unavailable.
    public static func movementEffort(
        steps: Int?,
        gravity: [GravitySample] = []
    ) -> Double? {
        if let steps, steps > 0 {
            return movementEffort(steps: steps)
        }
        guard let minutes = activeMotionMinutes(gravity), minutes > 0 else { return nil }
        return movementEffort(minutes: minutes)
    }

    /// Low-intensity movement floor from a measured/calibrated daily step total.
    public static func movementEffort(steps: Int) -> Double? {
        guard steps > 0 else { return nil }
        return movementEffort(minutes: Double(steps) / walkingCadenceStepsPerMinute)
    }

    /// Walking-level minutes represented by adjacent gravity records, or nil without usable evidence.
    public static func activeMotionMinutes(_ gravity: [GravitySample]) -> Double? {
        let rows = gravity
            .filter { $0.x.isFinite && $0.y.isFinite && $0.z.isFinite }
            .sorted { $0.ts < $1.ts }
        guard rows.count >= 2 else { return nil }

        var activeSeconds = 0.0
        for index in 1..<rows.count {
            let previous = rows[index - 1]
            let current = rows[index]
            let gap = Double(current.ts) - Double(previous.ts)
            guard gap > 0, gap <= maximumMotionEvidenceGapSeconds else { continue }

            let dx = current.x - previous.x
            let dy = current.y - previous.y
            let dz = current.z - previous.z
            let intensity = (dx * dx + dy * dy + dz * dz).squareRoot()
            if intensity >= walkingMotionThresholdG {
                activeSeconds += min(gap, maximumCreditedMotionSeconds)
            }
        }
        return activeSeconds > 0 ? activeSeconds / 60 : nil
    }

    private static func movementEffort(minutes: Double) -> Double? {
        guard minutes.isFinite, minutes > 0 else { return nil }
        let lowIntensityTRIMP = minutes * lightMovementWeight
        return min(
            StrainScorer.maxStrain,
            StrainScorer.trimpToStrain(lowIntensityTRIMP)
        )
    }
}

File: android/app/src/main/java/com/noop/analytics/DailyEffortScorer.kt
package com.noop.analytics

import com.noop.data.GravitySample
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Resolves one daily Effort value from cardiovascular load and low-intensity movement.
 *
 * Edwards TRIMP contributes nothing below 50% heart-rate reserve. For the daily score, measured steps
 * therefore provide a conservative movement floor: steps become walking minutes at 100 steps/minute,
 * each minute receives one quarter of Edwards zone 1's weight, and the resulting low-intensity TRIMP
 * uses the same logarithmic 0..100 map. The final score is max(cardio, movement), never a sum, so workout
 * steps are not counted twice.
 *
 * When a step counter is unavailable, dense-enough wrist gravity provides a conservative fallback.
 * Missing movement leaves the cardiovascular score unchanged. This is approximate, not a reproduction
 * of any proprietary wearable score. Mirrors DailyEffortScorer.swift.
 */
object DailyEffortScorer {
    const val walkingCadenceStepsPerMinute: Double = 100.0
    const val lightMovementWeight: Double = 0.25
    const val walkingMotionThresholdG: Double = 0.15
    const val maximumMotionEvidenceGapSeconds: Double = 120.0
    const val maximumCreditedMotionSeconds: Double = 60.0

    fun score(
        cardioEffort: Double?,
        steps: Int?,
        gravity: List<GravitySample> = emptyList(),
    ): Double? {
        val movement = movementEffort(steps, gravity) ?: return cardioEffort
        return cardioEffort?.let { maxOf(it, movement) } ?: movement
    }

    fun movementEffort(
        steps: Int?,
        gravity: List<GravitySample> = emptyList(),
    ): Double? {
        if (steps != null && steps > 0) return movementEffort(steps)
        val minutes = activeMotionMinutes(gravity) ?: return null
        return movementEffortFromMinutes(minutes)
    }

    fun movementEffort(steps: Int): Double? {
        if (steps <= 0) return null
        return movementEffortFromMinutes(steps.toDouble() / walkingCadenceStepsPerMinute)
    }

    fun activeMotionMinutes(gravity: List<GravitySample>): Double? {
        val rows = gravity
            .filter { it.x.isFinite() && it.y.isFinite() && it.z.isFinite() }
            .sortedBy { it.ts }
        if (rows.size < 2) return null

        var activeSeconds = 0.0
        for (index in 1 until rows.size) {
            val previous = rows[index - 1]
            val current = rows[index]
            val gap = current.ts.toDouble() - previous.ts.toDouble()
            if (gap <= 0.0 || gap > maximumMotionEvidenceGapSeconds) continue

            val dx = current.x - previous.x
            val dy = current.y - previous.y
            val dz = current.z - previous.z
            val intensity = sqrt(dx * dx + dy * dy + dz * dz)
            if (intensity >= walkingMotionThresholdG) {
                activeSeconds += min(gap, maximumCreditedMotionSeconds)
            }
        }
        return if (activeSeconds > 0.0) activeSeconds / 60.0 else null
    }

    private fun movementEffortFromMinutes(minutes: Double): Double? {
        if (!minutes.isFinite() || minutes <= 0.0) return null
        val lowIntensityTrimp = minutes * lightMovementWeight
        return min(StrainScorer.maxStrain, StrainScorer.trimpToStrain(lowIntensityTrimp))
    }
}
